### PR TITLE
Turn number based protocol

### PR DESCRIPTION
**Pull request by andrewgordstewart**
This PR changes the requirements on `forceMove`, `checkpoint` and `conclude`:
1. They all still require that the channel is not finalized.
2. The caller no longer has to know the correct current fingerprint. Instead, if given a supported state with a large enough turn number, the transaction succeeds.

(1) is somewhat different from the current spec, which states that `forceMove` is only allowable when the channel is open. I am not sure what behaviour is better. Either choice should give the same safety guarantees, so I think the choice should be driven by the user experience.

Because of (2), `concludeFromOpen` and `concludeFromChallenge` are no longer needed, and `_conclude` is made public.

Incidentally, this change generally seems more gas-efficient, though that may be due to more test cases supplying a single supported state.

Master, from [here](https://circleci.com/gh/statechannels/nitro-protocol/403?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link): 
```
Method Name                        Calls  Min Gas  Max Gas  Average Gas

setChannelStorageHash              36     29198    46310    42559      
forceMove                          7      63811    183386   113140     
checkpoint                         8      78838    138805   106989     
concludeFromChallenge              4      64863    92972    79175      
concludeFromOpen                   7      58065    98852    73000 
refute                             5      52574    73718    62901      
respond                            5      59987    85754    69714            
```

This branch, from [here](https://circleci.com/gh/statechannels/nitro-protocol/439):
```
Method Name                        Calls  Min Gas  Max Gas  Average Gas

setChannelStorageHash              44     29240    45840    42331      
forceMove                          12     64000    183601   112771     
checkpoint                         10     58800    107117   83630      
conclude                           11     56556    98775    75368      
refute                             6      53976    74624    63593      
respond                            5      61079    86225    70227        
```

### CODE REVIEW

File: docs/adjudicator/channel-storage.md [hunk 1]
```diff
@@ -3,74 +3,59 @@ id: channel-storage
...
 ```solidity
     mapping(bytes32 => bytes32) public channelStorageHashes;
 ```
 
+The `fingerprint` uniquely identifies the channel's current state, up to hash collisions.
+
 ## Turn number record
 
-`turNumRecord` is the highest turn number that is known to the chain to be supported by a full set of signatures. For example, the `turnNumRecord` might be increased by a submitted transaction including
+`turNumRecord` is the highest turn number that is known to the chain to be supported by a full set of signatures.
```
Comment (andrewgordstewart) [inline docs/adjudicator/channel-storage.md]: This is now true!

File: test/contracts/ForceMove/forceMove.test.ts [hunk 2]
```diff
@@ -51,96 +54,110 @@ beforeAll(async () => {
...
+        challengeDuration,
+        outcome,
+        appDefinition,
+        appData: defaultAbiCoder.encode(['uint256'], [data]),
+      }));
       const variableParts = states.map(state => getVariablePart(state));
       const fixedPart = getFixedPart(states[0]);
 
       // sign the states
       // sign the states
```
Comment (andrewgordstewart) [inline test/contracts/ForceMove/forceMove.test.ts]:
```suggestion
```

File: test/contracts/ForceMove/forceMove.test.ts [hunk 3]
```diff
@@ -51,96 +54,110 @@ beforeAll(async () => {
...
-
-        await expectRevert(() => {
-          return sendTransaction(provider, ForceMove.address, transactionRequest);
-        }, regex);
+        await expectRevert(() => tx, reasonString);
       } else {
         challengeRegisteredEvent = newChallengeRegisteredEvent(ForceMove, channelId);
 
-        await sendTransaction(provider, ForceMove.address, transactionRequest);
+        await tx;
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply test/contracts/ForceMove/forceMove.test.ts]: It's explained in the [commit](https://github.com/statechannels/nitro-protocol/commit/f3769aa5c421900e3ddc13877e10ed2b758ec61a).

Why is (i) preferable over (ii)?

[the remark this answers is not in the document]
Comment (geoknee) [reply test/contracts/ForceMove/forceMove.test.ts]: Makes sense 👍. I don't necessarily think (i) is preferable. I suppose (i) integrates the transaction creator into the test, too.

File: contracts/ForceMove.sol [hunk 4]
```diff
@@ -511,8 +423,9 @@ contract ForceMove {
         return stateHashes[stateHashes.length - 1];
     }
 
-    function _validTransitionChain(
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply contracts/ForceMove.sol]: That was just a typo, actually.

**Pull request merged**